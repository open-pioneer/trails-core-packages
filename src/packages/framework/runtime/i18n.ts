import { createIntl, createIntlCache, IntlFormatters, IntlShape } from "@formatjs/intl";
import { ErrorId } from "./errors";
import { createLogger, Error } from "@open-pioneer/core";
import { ApplicationMetadata } from "./metadata";
const LOG = createLogger("runtime:i18n");

/**
 * Represents i18n info for the entire application.
 * Currently not exposed to user code.
 */
export interface AppI18n {
    /** Chosen locale */
    readonly locale: string;

    /** Supported locales from app metadata. */
    readonly supportedLocales: string[];

    /** Given the package name, constructs a package i18n instance. */
    createPackageI18n(packageName: string): PackageIntl;
}

/**
 * Gives access to the package's i18n messages for the current locale.
 *
 * See also https://formatjs.io/docs/intl
 */
export type PackageIntl = Pick<IntlShape, "locale" | "timeZone"> & IntlFormatters<string>;

function createPackageIntl(locale: string, messages: Record<string, string>) {
    const cache = createIntlCache();
    return createIntl(
        {
            locale,
            messages
        },
        cache
    );
}

/**
 * Initializes the application's locale and fetches the appropriate i18n messages.
 */
export async function initI18n(
    appMetadata: ApplicationMetadata | undefined,
    forcedLocale: string | undefined
): Promise<AppI18n> {
    const supportedLocales = appMetadata?.locales ?? [];
    const userLocales = getBrowserLocales();
    if (LOG.isDebug()) {
        LOG.debug(
            `Attempting to pick locale for user (locales: ${userLocales.join(
                ", "
            )}) from app (supported locales: ${supportedLocales.join(
                ", "
            )}) [forcedLocale=${forcedLocale}].`
        );
    }

    const { locale, messageLocale } = pickLocale(
        forcedLocale,
        supportedLocales,
        getBrowserLocales()
    );

    if (LOG.isDebug()) {
        LOG.debug(`Using locale '${locale}' with messages from locale '${messageLocale}'.`);
    }

    let messages: Record<string, Record<string, string>>;
    if (supportedLocales.includes(messageLocale)) {
        try {
            messages = (await appMetadata?.loadMessages?.(messageLocale)) ?? {};
        } catch (e) {
            throw new Error(
                ErrorId.INTERNAL,
                `Failed to load messages for locale '${messageLocale}'.`,
                {
                    cause: e
                }
            );
        }
    }
    return {
        locale,
        supportedLocales,
        createPackageI18n(packageName) {
            const packageMessage = messages?.[packageName] ?? {};
            return createPackageIntl(locale, packageMessage);
        }
    };
}

/** Creates an empty i18n instance, e.g. for tests. */
export function createEmptyI18n(locale = "en"): PackageIntl {
    return createPackageIntl(locale, {});
}

export interface LocalePickResult {
    /** The actual locale (e.g. en-US) for number and date formatting etc. */
    locale: string;

    /**
     * The locale identifier to load messages for. E.g. "en".
     */
    messageLocale: string;
}

/**
 * Picks a locale for the app. Exported for tests.
 *
 * @param forcedLocale Optional forced locale (must be satisfied)
 * @param appLocales Locales the app has i18n messages for (e.g. "en", "de")
 * @param userLocales Locales requested by the user's browser
 */
export function pickLocale(
    forcedLocale: string | undefined,
    appLocales: string[],
    userLocales: string[]
): LocalePickResult {
    const pickImpl = (requestedLocales: string[]): LocalePickResult | undefined => {
        for (const requestedLocale of requestedLocales) {
            // try exact match
            if (appLocales.includes(requestedLocale)) {
                return { messageLocale: requestedLocale, locale: requestedLocale };
            }

            // try plain language tag (e.g. "en") as fallback
            const plainLanguage = requestedLocale.match(/^([a-z]+)/i)?.[1];
            if (plainLanguage && appLocales.includes(plainLanguage)) {
                return { messageLocale: plainLanguage, locale: requestedLocale };
            }
        }
        return undefined;
    };

    // Attempt to satisfy forced locale
    if (forcedLocale) {
        const result = pickImpl([forcedLocale]);
        if (!result) {
            const localesList = appLocales.join(", ");
            throw new Error(
                ErrorId.UNSUPPORTED_LOCALE,
                `Locale '${forcedLocale}' cannot be forced because it is not supported by the application.` +
                    ` Supported locales are ${localesList}.`
            );
        }
        if (result.locale !== result.messageLocale) {
            LOG.warn(
                `Non-exact match for forced locale '${forcedLocale}': using messages from '${result.messageLocale}'.`
            );
        }
        return result;
    }

    // Match preferred locale
    const supportedLocale = pickImpl(userLocales);
    if (supportedLocale) {
        return supportedLocale;
    }

    // Fallback: Most preferred locale (for dates etc.), but some of our messages
    return { locale: userLocales[0] ?? "en", messageLocale: appLocales[0] ?? "en" };
}

/**
 * Returns locales supported by the browser, in order of preference (preferred first).
 *
 * See also https://developer.mozilla.org/en-US/docs/Web/API/Navigator/languages
 */
function getBrowserLocales(): string[] {
    if (window.navigator.languages && window.navigator.languages.length) {
        return Array.from(window.navigator.languages);
    }
    return [window.navigator.language];
}
