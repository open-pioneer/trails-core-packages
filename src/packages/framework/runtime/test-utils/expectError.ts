export function expectError(impl: () => unknown): Error {
    try {
        impl();
        throw new Error("expected error!");
    } catch (e) {
        if (e instanceof Error) {
            return e;
        }
        throw new Error("unexpected error value, not an instance of Error");
    }
}

export function expectAsyncError(impl: () => Promise<unknown>): Promise<Error> {
    const promise = impl();
    return promise.then(
        () => {
            throw new Error("expected error!");
        },
        (e) => {
            if (e instanceof Error) {
                return e;
            }
            throw new Error("unexpected error value, not an instance of Error");
        }
    );
}
