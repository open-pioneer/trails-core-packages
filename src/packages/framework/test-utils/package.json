{
    "name": "@open-pioneer/test-utils",
    "private": true,
    "dependencies": {
        "@testing-library/dom": "^9.0.0",
        "@formatjs/intl": "^2.6.7"
    },
    "peerDependencies": {
        "@open-pioneer/runtime": "workspace:^"
    },
    "devDependencies": {
        "@open-pioneer/runtime": "workspace:^",
        "@testing-library/react": "^14.0.0",
        "react": "^18.2.0",
        "react-dom": "^18.2.0"
    }
}
