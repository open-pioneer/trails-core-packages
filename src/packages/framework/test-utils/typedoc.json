{
    "extends": ["../../../../typedoc.base.json"],
    "entryPoints": ["./web-components.ts", "./react.tsx", "./services.ts"]
}
