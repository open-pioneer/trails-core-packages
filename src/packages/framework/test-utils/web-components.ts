import { waitFor, within } from "@testing-library/dom";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const GLOBAL = global as any;

function nextId(): number {
    GLOBAL["__WEB_COMPONENT_COUNT__"] ??= 0;
    const id = (GLOBAL["__WEB_COMPONENT_COUNT__"] += 1);
    return id;
}

/**
 * Renders the given component into the DOM and returns the new node.
 */
export async function renderComponent(
    component: CustomElementConstructor | string,
    options?: {
        container?: HTMLElement;
        attributes?: Record<string, string>;
    }
) {
    const tag = typeof component === "string" ? component : defineComponent(component);
    const container = options?.container ?? document.body;
    const node = container.ownerDocument.createElement(tag);
    for (const [key, value] of Object.entries(options?.attributes ?? {})) {
        node.setAttribute(key, value);
    }
    container.innerHTML = "";
    container.appendChild(node);
    return {
        node
    };
}

/**
 * Renders the given component into the DOM (via {@link renderComponent}) and works on its shadow root.
 * In order to use this function, the shadow root should be attached as `open` from inside the component class.
 *
 * After the shadow root has been accessed, the function searches (and waits) for an inner root container (by default `.pioneer-root`)
 * where all other searches should be executed.
 *
 * Returns the shadow root, the inner container, and a bound queries object that automatically searches in the inner container
 * instead of the whole document.
 */
export async function renderComponentShadowDOM(
    component: CustomElementConstructor | string,
    options?: {
        container?: HTMLElement;
        attributes?: Record<string, string>;
        innerContainerSelector?: string;
    }
) {
    const { node } = await renderComponent(component, options);
    const selector = options?.innerContainerSelector ?? ".pioneer-root";
    const { shadowRoot, innerContainer } = await waitFor(() => {
        const shadowRoot = node.shadowRoot;
        if (!shadowRoot) {
            throw new Error("Shadow root not attached.");
        }
        const innerContainer = shadowRoot.querySelector<HTMLElement>(selector);
        if (!innerContainer) {
            throw new Error("Inner container not ready.");
        }
        return { shadowRoot, innerContainer };
    });

    const queries = within(innerContainer);
    return {
        node,
        shadowRoot,
        innerContainer,
        queries
    };
}

/**
 * Defines the given component and returns the unique tag name.
 * This helper is necessary because tag names must not collide in the global elements registry (and
 * un-registration is not possible).
 */
export function defineComponent(
    component: CustomElementConstructor,
    options?: {
        nameHint?: string;
    }
) {
    const nameHint = options?.nameHint ?? "test-component";
    const tag = `${nameHint}-${nextId()}`;
    customElements.define(tag, component);
    return tag;
}
