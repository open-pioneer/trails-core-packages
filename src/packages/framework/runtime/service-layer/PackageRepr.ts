// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
import { Error } from "@open-pioneer/core";
import { ApplicationProperties } from "../CustomElement";
import { ErrorId } from "../errors";
import { PackageIntl, AppI18n } from "../i18n";
import { PackageMetadata, PropertyMetadata } from "../metadata";
import { parseReferenceSpec, ReferenceSpec } from "./InterfaceSpec";
import { ServiceRepr } from "./ServiceRepr";

export interface PackageReprOptions {
    name: string;
    intl: PackageIntl;
    services?: ServiceRepr[];
    uiReferences?: ReferenceSpec[];
    properties?: Record<string, unknown>;
}

export class PackageRepr {
    static create(
        data: PackageMetadata,
        i18n: PackageIntl,
        customProperties?: Record<string, unknown>
    ): PackageRepr {
        const name = data.name;
        const properties = data.properties ?? {};
        const finalProperties = customizeProperties(name, properties, customProperties);
        const services = Object.entries(data.services ?? {}).map<ServiceRepr>(
            ([name, serviceData]) => {
                if (name !== serviceData.name) {
                    throw new Error(
                        ErrorId.INVALID_METADATA,
                        "Invalid metadata: service name mismatch."
                    );
                }
                return ServiceRepr.create(data.name, serviceData, i18n, finalProperties);
            }
        );
        const uiReferences = data.ui?.references?.map((ref) => parseReferenceSpec(ref)) ?? [];
        return new PackageRepr({
            name,
            services,
            uiReferences,
            properties: finalProperties,
            intl: i18n
        });
    }

    /** Package name */
    readonly name: string;

    /** Services defined by the package */
    readonly services: readonly ServiceRepr[];

    /** Interfaces required by UI components. */
    readonly uiReferences: readonly Readonly<ReferenceSpec>[];

    /** Resolved (perhaps customized) package properties. */
    readonly properties: Readonly<Record<string, unknown>>;

    /** Locale-dependant i18n messages. */
    readonly intl: PackageIntl;

    constructor(options: PackageReprOptions) {
        const name = options.name;
        if (!isValidPackageName(name)) {
            throw new Error(ErrorId.INTERNAL, `Invalid package name: '${name}'.`);
        }

        this.name = name;
        this.services = options.services ?? [];
        this.uiReferences = options.uiReferences ?? [];
        this.properties = options.properties ?? {};
        this.intl = options.intl;
    }
}

export function createPackages(
    packages: Record<string, PackageMetadata>,
    i18n: AppI18n,
    customProperties?: ApplicationProperties
) {
    return Object.entries(packages).map<PackageRepr>(([name, packageMetadata]) => {
        if (name !== packageMetadata.name) {
            throw new Error(ErrorId.INVALID_METADATA, "Invalid metadata: package name mismatch.");
        }

        const packageI18n = i18n.createPackageI18n(name);
        return PackageRepr.create(packageMetadata, packageI18n, customProperties?.[name]);
    });
}

/**
 * Merges the declared properties of the package with the application-defined
 * custom properties.
 */
function customizeProperties(
    packageName: string,
    properties: Record<string, PropertyMetadata>,
    customProperties: Record<string, unknown> = {}
) {
    const merged: Record<string, unknown> = {};
    const required: string[] = [];
    for (const [key, propertyMetadata] of Object.entries(properties)) {
        merged[key] = propertyMetadata.value;
        if (propertyMetadata.required) {
            required.push(key);
        }
    }

    for (const [key, value] of Object.entries(customProperties)) {
        if (!hasProperty(merged, key)) {
            throw new Error(
                ErrorId.INVALID_PROPERTY_NAME,
                `Unexpected property name '${key}' for package '${packageName}': the property does not exist.`
            );
        }
        merged[key] = value;
    }

    for (const key of required) {
        const value = merged[key];
        if (value == null) {
            throw new Error(
                ErrorId.REQUIRED_PROPERTY,
                `Package '${packageName}' requires the property '${key}' to be initialized to a non-null value.`
            );
        }
    }

    return merged;
}

const HAS_PROP = Object.prototype.hasOwnProperty;

function hasProperty(object: unknown, key: string) {
    return HAS_PROP.call(object, key);
}

// http://json.schemastore.org/package
const NAME_REGEX = /^(?:@[a-z0-9-*~][a-z0-9-*._~]*\/)?[a-z0-9-~][a-z0-9-._~]*$/;

function isValidPackageName(name: string): boolean {
    return NAME_REGEX.test(name);
}
