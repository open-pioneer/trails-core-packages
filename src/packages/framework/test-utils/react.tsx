import { createIntl, createIntlCache, IntlShape } from "@formatjs/intl";
import { Service } from "@open-pioneer/runtime";
import {
    PackageContext as InternalPackageContext,
    PackageContextMethods
} from "@open-pioneer/runtime/react-integration";
import { FC, ReactNode, useMemo } from "react";

export type AnyService = Service<Record<string, unknown>>;

export interface PackageContextProviderProps {
    /** Interface implementations, keyed by interface name. */
    services?: {
        [interfaceName: string]: AnyService;
    };

    /** Interface implementations, keyed by interface name and then by qualifier. */
    qualifiedServices?: {
        [interfaceName: string]: {
            [qualifier: string]: AnyService;
        };
    };

    /** Package properties (keyed by package name). */
    properties?: {
        [packageName: string]: Record<string, unknown>;
    };

    /**
     * The locale for i18n messages and formatting.
     *
     * @default "en"
     */
    locale?: string;

    /**
     * I18n messages for packages
     */
    messages?: {
        /**
         * I18n messages as (messageId, message) entries.
         *
         * @default {}
         */
        [packageName: string]: Record<string, string>;
    };

    /** Children to render */
    children?: ReactNode;
}

/**
 * Injects services and properties into the component tree.
 * React components using pioneer hooks like `useService` and `useProperties`
 * will receive the mocked properties here instead.
 */
export const PackageContextProvider: FC<PackageContextProviderProps> = (props) => {
    const { children, ...rest } = props;
    const contextMethods = useMemo(() => createPackageContextMethods(rest), [rest]);
    return (
        <InternalPackageContext.Provider value={contextMethods}>
            {children}
        </InternalPackageContext.Provider>
    );
};

function createPackageContextMethods(
    options: Omit<PackageContextProviderProps, "children">
): PackageContextMethods {
    const services = options?.services ?? {};
    const qualifiedServices = options?.qualifiedServices ?? {};
    const properties = options?.properties ?? {};
    const locale = options?.locale ?? "en";
    const messages = options?.messages ?? {};
    const cachedIntl: Record<string, IntlShape> = {};
    return {
        getService(packageName, interfaceName, options) {
            if (!options.qualifier) {
                const service = services[interfaceName];
                if (service) {
                    return service;
                }
                const qualified = qualifiedServices[interfaceName];
                if (qualified) {
                    for (const value of Object.values(qualified)) {
                        return value; // Return any implementation
                    }
                }
                throw new Error(
                    `Interface name not bound for testing: '${interfaceName}'. Update the configuration of PackageContextProvider.`
                );
            }

            const qualifiedService = qualifiedServices[interfaceName]?.[options.qualifier];
            if (!qualifiedService) {
                throw new Error(
                    `Interface name not bound for testing: '${interfaceName}' (qualifier '${options.qualifier}'). Update the configuration of PackageContextProvider.`
                );
            }
            return qualifiedService;
        },
        getServices(packageName, interfaceName) {
            const unqualified = services[interfaceName];
            const results = Object.values(qualifiedServices[interfaceName] ?? {});
            if (unqualified) {
                results.push(unqualified);
            }
            return results;
        },
        getProperties(packageName) {
            const packageProperties = properties[packageName];
            if (!packageProperties) {
                throw new Error(
                    `No properties for package '${packageName}' bound for testing. Update the configuration of PackageContextProvider.`
                );
            }
            return packageProperties;
        },
        getIntl(packageName) {
            const initIntl = () => {
                const packageMessages = messages[packageName];
                const cache = createIntlCache();
                return createIntl(
                    {
                        locale,
                        messages: packageMessages
                    },
                    cache
                );
            };
            return (cachedIntl[packageName] ??= initIntl());
        }
    };
}
