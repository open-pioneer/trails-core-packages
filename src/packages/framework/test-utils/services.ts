// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
import { ReferenceMeta, Service, ServiceConstructor } from "@open-pioneer/runtime";
import { createIntl, createIntlCache } from "@formatjs/intl";

/**
 * Options for the {@link createService} function.
 */
export interface CreateServiceOptions<References> {
    /**
     * References provided to the service constructor.
     *
     * @default {}
     */
    references?: PartialServiceReferences<References>;

    /**
     * Properties provided to the service constructor.
     *
     * @default {}
     */
    properties?: Record<string, unknown>;

    /**
     * The locale for i18n messages and formatting.
     *
     * @default "en"
     */
    locale?: string;

    /**
     * The locale for embedded default messages.
     *
     * See also https://formatjs.io/docs/intl#message-descriptor
     *
     * @default "en"
     */
    defaultMessageLocale?: string;

    /**
     * I18n messages as (messageId, message) entries.
     *
     * @default {}
     */
    messages?: Record<string, string>;
}

type PartialServiceReferences<References> = {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    [referenceName in keyof References]?: References[referenceName] extends any[]
        ? Partial<References[referenceName][number]>[]
        : Partial<References[referenceName]>;
};

/**
 * Creates a new service instance of `clazz`.
 *
 * The options passed into the constructor of `clazz` are taking from the `options` parameter.
 *
 * This function allows the user to define mock- or test objects as service references.
 *
 * Most type-checking features are disabled for convenience, but by default this function's types
 * enforce that a value provided for a reference must match the partial interface of the actually required type.
 *
 * For example, if a service depends on `{ ctx: ApplicationContext }`, this function ensures
 * that the test value for `ctx` implements `Partial<ApplicationContext>`.
 * This can of course be disabled by casting the value explicitly.
 *
 * @param clazz a service class
 * @param options options passed to the service constructor
 * @returns a new instance of the given service class
 */
export async function createService<References extends {}, Interface extends {}>(
    clazz: ServiceConstructor<References, Interface>,
    options?: CreateServiceOptions<References>
): Promise<Service<Interface>> {
    const references = options?.references ?? {};
    const referencesMeta = Object.fromEntries(
        Object.entries(references).map(([referenceName, reference]) => [
            referenceName,
            referenceMeta(referenceName, reference)
        ])
    );

    const locale = options?.locale ?? "en";
    const defaultMessageLocale = options?.defaultMessageLocale ?? "en";
    const messages = options?.messages ?? {};
    const cache = createIntlCache();
    const intl = createIntl(
        {
            locale,
            defaultLocale: defaultMessageLocale,
            messages
        },
        cache
    );

    return new clazz({
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        references: references as any,
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        referencesMeta: referencesMeta as any,
        properties: options?.properties ?? {},
        intl
    });
}

function referenceMeta(referenceName: string, reference: unknown): ReferenceMeta | ReferenceMeta[] {
    if (Array.isArray(reference)) {
        return reference.map<ReferenceMeta>((_, index) => ({
            serviceId: `test-utils::${referenceName}-${index}`
        }));
    }
    return {
        serviceId: `test-utils::${referenceName}`
    };
}
