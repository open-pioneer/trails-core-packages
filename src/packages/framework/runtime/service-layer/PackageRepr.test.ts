// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
import { expect, it } from "vitest";
import { AppI18n, createEmptyI18n, PackageIntl } from "../i18n";
import { PackageMetadata } from "../metadata";
import { expectError } from "../test-utils/expectError";
import { createPackages, PackageRepr } from "./PackageRepr";

class ClazzA {}

class ClazzB {}

it("parses package metadata into internal package representations", function () {
    const metadata: Record<string, PackageMetadata> = {
        a: {
            name: "a",
            services: {
                A: {
                    name: "A",
                    clazz: ClazzA,
                    provides: [],
                    references: {
                        foo: {
                            name: "b.ServiceB1"
                        }
                    }
                }
            }
        },
        b: {
            name: "b",
            services: {
                B: {
                    name: "B",
                    clazz: ClazzB,
                    provides: [
                        {
                            name: "b.ServiceB1"
                        },
                        {
                            name: "b.ServiceB2"
                        }
                    ],
                    references: {}
                }
            },
            properties: {
                foo: {
                    value: 123,
                    required: false
                }
            }
        }
    };

    const testi18n: AppI18n = {
        locale: "test-locale",
        supportedLocales: [],
        createPackageI18n() {
            return createEmptyI18n("zh-CN");
        }
    };

    const packages = createPackages(metadata, testi18n);
    expect(packages).toHaveLength(2);

    const packageA = packages.find((b) => b.name === "a")!;
    expect(packageA).toBeDefined();
    expect(packageA.intl.locale).toEqual("zh-CN");

    const serviceA = packageA.services.find((s) => s.name === "A")!;
    expect(serviceA).toBeDefined();
    expect(serviceA.id).toStrictEqual("a::A");
    expect(serviceA.state).toStrictEqual("not-constructed");
    expect(serviceA.intl.locale).toEqual("zh-CN");
    expect(serviceA.instance).toBeUndefined();
    expect(serviceA.dependencies).toStrictEqual([
        {
            referenceName: "foo",
            interfaceName: "b.ServiceB1",
            qualifier: undefined
        }
    ]);
    expect(serviceA.interfaces).toStrictEqual([]);

    const packageB = packages.find((b) => b.name === "b")!;
    expect(packageB).toBeDefined();
    expect(packageB.intl.locale).toEqual("zh-CN");

    const serviceB = packageB.services.find((s) => s.name === "B")!;
    expect(serviceB).toBeDefined();
    expect(serviceB.interfaces).toStrictEqual([
        { interfaceName: "b.ServiceB1", qualifier: undefined },
        { interfaceName: "b.ServiceB2", qualifier: undefined }
    ]);
    expect(serviceB.properties.foo).toBe(123);
});

it("supports package properties", function () {
    const metadata: PackageMetadata = {
        name: "foo",
        properties: {
            propertyName: {
                value: "propertyValue"
            }
        }
    };
    const pkg = createPackageFromMetadata(metadata);
    expect(pkg.properties).toStrictEqual({
        propertyName: "propertyValue"
    });
});

it("supports package properties in services", function () {
    const metadata: PackageMetadata = {
        name: "foo",
        properties: {
            propertyName: {
                value: "propertyValue"
            }
        },
        services: {
            ServiceA: {
                name: "ServiceA",
                clazz: class {}
            }
        }
    };
    const pkg = createPackageFromMetadata(metadata);
    const service = pkg.services[0]!;
    expect(service).toBeDefined();
    expect(service.properties).toStrictEqual({
        propertyName: "propertyValue"
    });
});

it("supports customization of package properties", function () {
    const metadata: PackageMetadata = {
        name: "foo",
        properties: {
            propertyName: {
                value: "propertyValue"
            }
        }
    };
    const pkg = createPackageFromMetadata(metadata, {
        propertyName: "override"
    });
    expect(pkg.properties).toStrictEqual({
        propertyName: "override"
    });
});

it("throws an error if a required property is not specified", function () {
    const metadata: PackageMetadata = {
        name: "foo",
        properties: {
            propertyName: {
                value: null,
                required: true
            }
        }
    };

    const err = expectError(() => createPackageFromMetadata(metadata));
    expect(err.message).toMatchSnapshot();
});

it("throws no error if a required property is specified", function () {
    const metadata: PackageMetadata = {
        name: "foo",
        properties: {
            propertyName: {
                value: null,
                required: true
            }
        }
    };
    const pkg = createPackageFromMetadata(metadata, {
        propertyName: 123
    });
    expect(pkg).toBeDefined();
});

it("throws an error if an attempt is made to customize a non-existent property", function () {
    const metadata: PackageMetadata = {
        name: "foo",
        properties: {
            propertyName: {
                value: null
            }
        }
    };

    const err = expectError(() =>
        createPackageFromMetadata(metadata, {
            otherProperty: 123
        })
    );
    expect(err.message).toMatchSnapshot();
});

it("passes package properties to created ServiceRepr instances", function () {
    const metadata: PackageMetadata = {
        name: "foo",
        properties: {
            propertyName: {
                value: null
            }
        },
        services: {
            Service: {
                name: "Service",
                clazz: class {}
            }
        }
    };
    const pkg = createPackageFromMetadata(metadata, {
        propertyName: 123
    });
    const service = pkg.services[0]!;
    expect(service.properties).toStrictEqual({
        propertyName: 123
    });
});

function createPackageFromMetadata(
    data: PackageMetadata,
    properties?: Record<string, unknown>,
    i18n?: PackageIntl
) {
    return PackageRepr.create(data, i18n ?? createEmptyI18n(), properties);
}
