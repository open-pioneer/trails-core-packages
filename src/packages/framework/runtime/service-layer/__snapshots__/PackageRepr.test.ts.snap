// Vitest Snapshot v1

exports[`throws an error if a required property is not specified 1`] = `"runtime:required-property: Package 'foo' requires the property 'propertyName' to be initialized to a non-null value."`;

exports[`throws an error if an attempt is made to customize a non-existent property 1`] = `"runtime:invalid-property-name: Unexpected property name 'otherProperty' for package 'foo': the property does not exist."`;
